From a1d1e8f5d9f4f267d5466f4217f34e734edee6e1 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Tue, 27 Sep 2016 10:58:29 -0700
Subject: [PATCH] DOCKER-939 docker build - symlink outside of build context
 caused crash Reviewed by: Julien Gilli <julien.gilli@joyent.com> Reviewed by:
 Trent Mick <trentm@gmail.com>

---
 lib/build.js       | 53 +++++++++++++++++++++++++++++++++++++++++----
 package.json       |  5 +++--
 test/test_build.js | 54 ++++++++++++++++++++++++++++++++++++++++++++--
 3 files changed, 104 insertions(+), 8 deletions(-)

diff --git a/lib/build.js b/lib/build.js
index 9b39026..15784a6 100644
--- a/lib/build.js
+++ b/lib/build.js
@@ -160,7 +160,6 @@ function Builder(opts) {
     // Generated image layers during build, each entry is map of:
     //   { cmd: Object, image: Object }
     this.layers = [];
-    this.realpathCache = {}; // Used to cache realpath lookups.
     this.stepNo = -1; // Command step number.
     this.totalNumSteps = 0;  // Number of dockerfile commands to be run.
     this.cmdSet = false;     // If a CMD entry has been processed.
@@ -1258,7 +1257,17 @@ function copyInfo_contextPath()
     var src = path.join(this.builder.contextExtractDir, this.origPath);
 
     // Sanity check that path is still inside the context extract dir.
-    src = fs.realpathSync(src, this.builder.realpathCache);
+    try {
+        src = fs.realpathSync(src);
+    } catch (e) {
+        if (e.code !== 'ENOENT') {
+            throw e;
+        }
+        // The only case where this realpath should fail, is to a symlink that
+        // does not exist.
+        assert.ok(fs.lstatSync(src).isSymbolicLink(),
+            'failed realpath must be a symlink');
+    }
     var extDirWithSlash = this.builder.contextExtractDir + '/';
 
     var pathOk = (src.substr(0, extDirWithSlash.length) === extDirWithSlash)
@@ -1292,7 +1301,37 @@ function copyInfo_basename()
 lazyProperty(copyInfo.prototype, 'stat',
 function copyInfo_stat()
 {
-    return fs.statSync(this.contextPath);
+    try {
+        return fs.statSync(this.contextPath);
+    } catch (e) {
+        if (e.code !== 'ENOENT') {
+            throw e;
+        }
+        // Allow symlink to a missing file.
+        var lstat = fs.lstatSync(this.contextPath);
+        if (!lstat.isSymbolicLink()) {
+            throw e;
+        }
+        return lstat;
+    }
+});
+
+// Return true if the contextPath is a *broken* symlink.
+lazyProperty(copyInfo.prototype, 'contextPathIsBrokenSymlink',
+function copyInfo_contextPathIsBrokenSymlink()
+{
+    // Note that the only way that stat.isSymbolicLink returns true, is if this
+    // is in fact a lstat for a broken symlink, see the 'stat' code for more
+    // detail.
+    return this.stat.isSymbolicLink();
+});
+
+// Return the linkname for this broken symlink.
+lazyProperty(copyInfo.prototype, 'contextPathBrokenSymlinkTarget',
+function copyInfo_contextPathBrokenSymlinkTarget()
+{
+    assert.ok(this.contextPathIsBrokenSymlink, 'Path must be a broken symlink');
+    return fs.readlinkSync(this.contextPath);
 });
 
 // Return true if the contextPath is a directory.
@@ -1334,6 +1373,11 @@ lazyProperty(copyInfo.prototype, 'checksum',
 function copyInfo_checksum()
 {
     assert.ok(!this.contextPathIsDirectory, '!contextPathIsDirectory');
+    if (this.contextPathIsBrokenSymlink) {
+        var hash = crypto.createHash('sha256');
+        hash.update('symlink:' + this.contextPathBrokenSymlinkTarget);
+        return hash.digest('hex');
+    }
     return utils.fileGetSha256Sync(this.contextPath);
 });
 
@@ -1929,7 +1973,8 @@ Builder.prototype.performCopy = function performCopy(cmd, copyInfos, callback) {
         // Special handling for ADD with a tar file.
         function detectAddTarFile(next) {
             if (cmd.name !== 'ADD' || copyInfos.length !== 1
-                || copyInfos[0].contextPathIsDirectory) {
+                || copyInfos[0].contextPathIsDirectory
+                || copyInfos[0].contextPathIsBrokenSymlink) {
 
                 next();
                 return;
diff --git a/package.json b/package.json
index 5bd984b..fcfa377 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,8 @@
 {
   "name": "sdc-docker-build",
-  "version": "0.1.0",
-  "description": "Parses a Dockerfile and returns an array of commands.",
+  "version": "0.1.1",
+  "description": "Docker build helper module for sdc-docker",
+  "private": true,
   "keywords": [
     "dockerfile",
     "docker",
diff --git a/test/test_build.js b/test/test_build.js
index 671433f..b23d337 100644
--- a/test/test_build.js
+++ b/test/test_build.js
@@ -186,6 +186,14 @@ function verifyFileContents(t, builder, filepath, contents) {
     }
 }
 
+function verifySymlink(t, builder, filepath, details) {
+    var fullpath = path.join(builder.containerRootDir, filepath);
+    if (fs.readlinkSync(fullpath) !== details.linkname) {
+        t.equal(fs.readlinkSync(fullpath), details.linkname,
+            'Link names for ' + filepath + ' do not match');
+    }
+}
+
 function verifyFilesystem(t, builder, containerPath, filesystem) {
     var entry;
     var fullpath = path.join(builder.containerRootDir, containerPath);
@@ -205,13 +213,16 @@ function verifyFilesystem(t, builder, containerPath, filesystem) {
         name = names[i];
         entry = filesystem[name];
         relPath = path.join(containerPath, name);
-        stat = fs.statSync(path.join(fullpath, name));
+        stat = fs.lstatSync(path.join(fullpath, name));
         if (stat.isDirectory()) {
             assert.object(entry, name);
             verifyFilesystem(t, builder, relPath, entry);
         } else if (stat.isFile()) {
             assert.string(entry, name);
             verifyFileContents(t, builder, relPath, entry);
+        } else if (stat.isSymbolicLink()) {
+            assert.object(entry, name);
+            verifySymlink(t, builder, relPath, entry);
         } else {
             t.fail('Unexpected file type at: ' + relPath);
         }
@@ -265,7 +276,11 @@ function createTarStream(fileAndContents) {
     var pack = tar.pack();
 
     Object.keys(fileAndContents).forEach(function (name) {
-        pack.entry({ name: name }, fileAndContents[name]);
+        if (typeof (fileAndContents[name]) === 'object') {
+            pack.entry(fileAndContents[name]);
+        } else {
+            pack.entry({ name: name }, fileAndContents[name]);
+        }
     });
 
     pack.finalize();
@@ -1261,6 +1276,41 @@ tape('symlinks', function (t) {
 });
 
 
+tape('symlinkMissing', function (t) {
+    var fileAndContents = {
+        'Dockerfile': [
+            'FROM busybox',
+            'ADD config /home/config/',
+            'ADD config/theMissingLink /myNewMissingLink'
+        ].join('\n'),
+        'config/theMissingLink': {
+            name: 'config/theMissingLink',
+            type: 'symlink',
+            linkname: '/missing/directory/path'
+        }
+    };
+
+    testBuildContents(t, fileAndContents, function (err, result) {
+        var builder = result.builder;
+        if (showError(t, err, builder)) {
+            return;
+        }
+
+        var expectedFilesystem = {
+            'home': {
+                'config': {
+                    'theMissingLink': fileAndContents['config/theMissingLink']
+                }
+            },
+            'myNewMissingLink': fileAndContents['config/theMissingLink']
+        };
+        verifyFilesystem(t, builder, '/', expectedFilesystem);
+
+        testEnd(t, builder);
+    });
+});
+
+
 tape('FROM must be first', function (t) {
     var fileAndContents = {
         'Dockerfile': 'MAINTAINER me\nFROM busybox\n'
-- 
2.21.0

